adding js option to track hostname

jvanasco opened the pull request:
This is a proposed feature to make the panel a little more useful when debugging multi-host applications.  

1. The JSON payload about a route is extended to include the hostname
2. A javascript toggle (default inactive) is offered to augment the `title` attribute of the sidebar elements with the hostname, which appears on mouseovers.

It has made my life a lot easier in a few situations.

stevepiercy commented:
I think that some docs are needed for this, specifically its usage and what it does. I think you could copy-paste what you have in `src/pyramid_debugtoolbar/templates/settings_tab.dbtmako`, and tweak it for reStructuredText.

jvanasco commented:
@stevepiercy I hope I didn't overdo things (or step on your toes too much), but I made a few changes/reorganization of the docs when implementing your requested docs addition:

* I added a new "Toolbar Settings" section, which deals with the navbar "settings". I migrated some docs from other places into this, but cross referenced it.  This was done to centralized and cross referenced a lot of the information on the "performance" panel settings.

* I renamed the "panel" headers for "Settings" and "Performance" to include the word "Panel".  I did this so the cross-referencing in the generated docs would be more clean and user friendly.

* The "Settings" panel and "Settings" tab now cross-reference one another and explain the difference.  This may sound confusing at first, but try looking through the generated docs and I think you'll be happy with the clarity this brings, as there are two "settings" sections on the toolbar.

* I did screencaps, then dropped them onto a "template" of an existing screencap so it all looks the same.

Thoughts?

stevepiercy commented on docs/index.rst:
```diff
@@ -386,11 +386,19 @@ version and platform itself.
+
+   The toolbar offers a `Settings` "tab" in the navigation bar
+   (see :ref:`toolbar_settings`), which is distinct from this `Settings`
+   "panel".
```
```suggestion
    For realtime customization of the toolbar and its panels, use the :guilabel:`Settings` *tab* in the navigation bar.
    See :ref:`toolbar_settings`.
```

stevepiercy replied:
meh, GitHub does not accept multi-line suggestions, but you get the idea.

stevepiercy commented on docs/index.rst:
```diff
@@ -386,11 +386,19 @@ version and platform itself.
-~~~~~~~~
+.. _panel_settings:
+
+Settings Panel
```
It is unnecessary to rename because it is already under the heading of "Toolbar Panels". Also it will blow out the left navigation and would be inconsistent with other panel subheadings.

```suggestion
Settings
```

stevepiercy commented on docs/index.rst:
```diff
@@ -386,11 +386,19 @@ version and platform itself.
+.. _panel_settings:
+
+Settings Panel
+~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,19 +430,25 @@ Displays messages logged by the current page.
-~~~~~~~~~~~
+.. _panel_performance:
+
+Performance Panel
```
ibid
```suggestion
Performance
```

stevepiercy commented on docs/index.rst:
```diff
@@ -422,19 +430,25 @@ Displays messages logged by the current page.
+.. _panel_performance:
+
+Performance Panel
+~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~
```

stevepiercy commented on docs/index.rst:
```diff
@@ -386,11 +386,19 @@ version and platform itself.
+Settings Panel
+~~~~~~~~~~~~~~
 
 Displays Pyramid deployment settings, i.e., ``registry.settings``.
```
```suggestion
Displays your Pyramid application's deployment settings, i.e., ``registry.settings``.
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +492,42 @@ introspection system (available in Pyramid 1.3+ only).
+Toolbar Settings
+----------------
+
+The `Settings` "tab" allows for realtime customization of the toolbar and its
```
```suggestion
The :guilabel:`Settings` tab allows for realtime customization of the toolbar and its panels.
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +492,42 @@ introspection system (available in Pyramid 1.3+ only).
+The `Settings` "tab" allows for realtime customization of the toolbar and its
+panels.
+
+.. note::
```
```suggestion
.. seealso::
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +492,42 @@ introspection system (available in Pyramid 1.3+ only).
+
+.. note::
+
+   This refers to the toolbar's `Settings` "tab" in the navigation bar, which is
```
To display your Pyramid application's settings, see :ref:`panel_settings`.

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +492,42 @@ introspection system (available in Pyramid 1.3+ only).
+   This refers to the toolbar's `Settings` "tab" in the navigation bar, which is
+   distinct from the `Settings` "panel" (see :ref:`panel_settings`).
+
+.. image:: toolbar_settings.png
```
Move this above the `seealso` revision.

jvanasco replied:
I'm not sure what this means. Do you mean just the image?

stevepiercy replied:
Yes. It feels associated to the `seealso` because it is immediately after it, and it should be associated with the text above that.

stevepiercy commented on docs/index.rst:
```diff
@@ -386,11 +386,19 @@ version and platform itself.
 
 Displays Pyramid deployment settings, i.e., ``registry.settings``.
 
+.. note::
```
```suggestion
.. seealso::
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +492,42 @@ introspection system (available in Pyramid 1.3+ only).
+An internal profiler can be enabled through the "performance" checkmark in the
+`Settings` tab in the navigation bar. When the checkbox is green, the request
+will be profiled and profiling information will be gathered and displayed on the
+`Performance` panel output.  (see :ref:`panel_performance`).
```
```suggestion
An internal profiler can be enabled under the :guilabel:`Settings` tab in the navigation bar.
Click the red :guilabel:`X` mark to enable the profiler.
When the mark is a green :guilabel:`check`, the request will be profiled and profiling information will be gathered and displayed on the :guilabel:`Performance` panel output.
See :ref:`panel_performance`.
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +492,42 @@ introspection system (available in Pyramid 1.3+ only).
+
+Some Pyramid applications will exist on multiple hosts. To aid in debugging
+these applications, you can enable the ``hostname`` to appear on mouseover
+titles of the "Request" selector column on the left side of the toolbar, instead
```
```suggestion
titles of the :guilabel:`Request` selector column on the left side of the toolbar, instead
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +492,42 @@ introspection system (available in Pyramid 1.3+ only).
+Some Pyramid applications will exist on multiple hosts. To aid in debugging
+these applications, you can enable the ``hostname`` to appear on mouseover
+titles of the "Request" selector column on the left side of the toolbar, instead
+of the path. This functionality is disabled by default; This setting will not
```
```suggestion
of the path. This functionality is disabled by default. This setting will not
```

jvanasco commented:
@stevepiercy The current state merges all your requests (except one i was unsure about) with a slight twist.  

    See :ref:`toolbar_settings_performance` [:guilabel:`Toolbar Settings`].

Is used to render this on HTML:

![Screen Shot 2020-01-09 at 4 01 39 PM](https://user-images.githubusercontent.com/204779/72104754-1abc9200-32fa-11ea-8bb8-663050c7e401.png)

   To display your Pyramid application's settings, see :ref:`panel_settings` [:guilabel:`Panels`].

Is used to render this on HTML:

![Screen Shot 2020-01-09 at 4 01 03 PM](https://user-images.githubusercontent.com/204779/72104813-3889f700-32fa-11ea-8f4f-50758de56aa7.png)

See :ref:`panel_performance` [:guilabel:`Panels`].

corresponds to:

![Screen Shot 2020-01-09 at 4 01 24 PM](https://user-images.githubusercontent.com/204779/72104864-52c3d500-32fa-11ea-9415-c4432e55470a.png)

This is to maintain clarity after dropping the repeated word 'Panel' and "Settings" from the subtitles. In the original commit, I used those words (at the expense of the TOC being streamlined) so we could just rely on the auto-generated text -- otherwise the see-also has the right links... but the text is nearly identical for the two "settings".

stevepiercy commented on docs/index.rst:
```diff
@@ -422,19 +429,24 @@ Displays messages logged by the current page.
-   displayed on the "Performance" panel output.
+#.  Under the :guilabel:`Settings` tab in the navigation bar, click the red :guilabel:`X` mark.
+    When there is a green :guilabel:`check` mark, each request will be profiled and profiling information will be gathered and displayed on the :guilabel:`Performance` panel output.
+    See :ref:`toolbar_settings_performance` [:guilabel:`Toolbar Settings`].
```
This is misuse of [`:guilabel:`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-guilabel). It is used to distinguish GUI elements from inline text. There is no GUI element with the label of "Toolbar Settings".
```suggestion
    See :ref:`Toolbar Settings <toolbar_settings_performance>`.
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +490,44 @@ introspection system (available in Pyramid 1.3+ only).
+----------------
+
+The :guilabel:`Settings` tab allows for realtime customization of the toolbar and its panels.
+panels.
```
This is a duplicate line.
line.

```suggestion
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +490,44 @@ introspection system (available in Pyramid 1.3+ only).
+
+.. seealso::
+
+   To display your Pyramid application's settings, see :ref:`panel_settings` [:guilabel:`Panels`].
```
Misuse of `guilabel`.

```suggestion
   To display your Pyramid application's settings, see :ref:`Settings panel <panel_settings>`.
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +490,44 @@ introspection system (available in Pyramid 1.3+ only).
+An internal profiler can be enabled under the :guilabel:`Settings` tab in the navigation bar.
+Click the red :guilabel:`X` mark to enable the profiler.
+When the mark is a green :guilabel:`check`, the request will be profiled and profiling information will be gathered and displayed on the :guilabel:`Performance` panel output.
+See :ref:`panel_performance` [:guilabel:`Panels`].
```
Misuse of `guilabel`.

```suggestion
See :ref:`Performance panel <panel_performance>`.
```

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +490,44 @@ introspection system (available in Pyramid 1.3+ only).
+
+An internal profiler can be enabled through the "performance" checkmark in the
+`Settings` tab in the navigation bar. When the checkbox is green, the request
+will be profiled and profiling information will be gathered and displayed on the
```
Remove obsolete duplicate lines.

stevepiercy commented on docs/index.rst:
```diff
@@ -478,6 +490,44 @@ introspection system (available in Pyramid 1.3+ only).
+
+Some Pyramid applications will exist on multiple hosts. To aid in debugging
+these applications, you can enable the ``hostname`` to appear on mouseover
+titles of the :guilabel:`Request` selector column on the left side of the toolbar, instead
```
Name correction.

```suggestion
titles of the :guilabel:`Requests` selector column on the left side of the toolbar, instead
```

stevepiercy commented:
I wonder if we should stop punching ourselves in the face, and rename the Toolbar "Settings" to "Toolbar Config" in the main navigation? It would be a lot of work, new screenshots, and probably not worth it. Just thinking out loud.

In all practicality, though, do people really get confused between the two "Settings" such that they are unable to use the toolbar? I do not recall seeing any issue reported for this, so I don't see the need and I think we're getting too deep in the weeds.

jvanasco commented:
I don't think people are currently confused, and I think the debugtoolbar is pretty intuitive itself. My only concern is when dealing with documentation, and being clear about which "Settings" (panel or toolbar) is being referred to in a given piece of text.

jvanasco commented:
Missed them. I'll get to this by Monday.

Sent from my iPhone

> On Jan 11, 2020, at 1:29 AM, Steve Piercy <notifications@github.com> wrote:
> 
> @stevepiercy requested changes on this pull request.
> 
> There's two more changes requested that have not been addressed. Did you miss them?
> 
> Once those are done, then this PR is good to merge. Thank you!
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jvanasco commented:
Wow, Steve. I am so sorry for putting you through this. I honestly don't know how I missed that.

stevepiercy commented:
OK, this passes my review. I've requested review from PyPI maintainers @mmerickel, @bertjwregeer, and @blaflamme for their final review, and a possible release to PyPI. Thank you!

mmerickel commented:
I'm failing to see why we don't just have this always on instead of configurable? There's been a lot of work put into this PR to make it configurable when the data is just a tiny addition to the tooltip that seems generally useful.

jvanasco commented:
The existing behavior (default in this PR) is to have the title mouseover to the full path.  My personal experience: this mouseover is often the main differentiator between requests with longer paths. 

I did not want to lose that utility. While bootstrap has a component that offers multi-line mouseovers, which could support both at once, it's not used elsewhere in the toolbar and would be a bit of a change to the UX.

mmerickel commented:
what about changing the tooltip to show the full request url?

jvanasco commented:
that's a great idea. i'll give it a try!

jvanasco commented:
@mmerickel here's a version that just does host+path ; I did not include the scheme

@stevepiercy i updated the docs to reflect this change. i did find one small issue we missed before, which i'll do an inline comment for.

jvanasco commented on docs/index.rst:
```diff
@@ -343,7 +343,7 @@ Most panels do not support any extra active features and need not be
 :attr:`~pyramid_debugtoolbar.panel.DebugPanel.is_active` state in which they
 will do some extra work. For example, the
-:class:`~pyramid_debugtoolbar.panels.performance.PerformanceDebugPanel``
+:ref:`PerformanceDebugPanel <panel_performance>`
```
The autolink for this should have been:

    :class:`~pyramid_debugtoolbar.panels.performance.PerformanceDebugPanel`

however after correcting that, it linked to the api docs which do nothing. I changed this link to read the same, but now link to the panel docs.

stevepiercy replied:
LGTM.

mmerickel commented:
Great, much simpler!
